### PR TITLE
V0.5 credential loading cleanup

### PR DESCRIPTION
**Pull request by erikerikson**
DELTA:
1. Make the logic of credential extraction from the configuration, environment (raw), or environment (via profile) very explicit and clear. Use built in classes as appropriate. (actually, we're not doing that for loading the ~/.aws/credentials file [via http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SharedIniFileCredentials.html] but maybe we should - going for minimum change)
2. Maintain optionality (but support for) session tokens.
3. Cover all possible environment prefixes consistently.

ORIGINAL:
This PR enables the use of SessionTokens when they are available.

Add AWS session token usage and handling (for NodeJs or Python). This is important to enable the use of temporary credentials, particularly for cases of credential federation (e.g. SAML) where end users are unable to establish permanent IAM credentials.

Credentials are obtained as follows
1. Will use ~/.aws/credentials defined credentials if an environment profile is set
2. Checks for environmentally defined credentials
3. Reads any duplicated credentials in admin.env.

Environment profiles/credentials obtained from the following environmental prefixes in listed order:
AWS
SERVERLESS_ADMIN_AWS
AWS_[STAGE]
SERVERLESS_ADMIN_AWS_[STAGE]

The logic for obtaining credentials and preparing AWS SDK config has been consolidated. As appropriate, where defaults are inappropriate, customization of the base default is necessary/implemented.

This PR should address the following list of issues:
#516
#488

### CODE REVIEW

**Comment by erikerikson** [comment]
@ac360 Third time is the charm?

File: lib/ProviderAws.js [hunk 1]
```diff
@@ -91,86 +91,79 @@ class ServerlessProviderAws {
...
-      };
-    }
-
-    if (!credentials) {
+    // implicitly already in the config...
+    this.addConfigurationCredentials(credentials, this._S.config);                       // use the given configuration credentials if they are the only available credentials.
+    // first from environment
+    this.addEnvironmentCredentials(credentials, 'AWS');                      // allow for Amazon standard credential environment variable prefix.
+    this.addEnvironmentCredentials(credentials, 'SERVERLESS_ADMIN_AWS');          // but override with more specific credentials if these are also provided.
+    this.addEnvironmentCredentials(credentials, 'AWS' + stage);                  // and also override these with the Amazon standard *stage specific* credential environment variable prefix.
```
Comment (hauboldj) [inline lib/ProviderAws.js]: I think this should be `this.addEnvironmentCredentials(credentials, 'AWS_' + stage);` (Missing underscore), same on the next line.

**Comment by hauboldj** [comment]
:+1: I really appreciate the support for different credentials for different stages.

File: lib/ProviderAws.js (hunk 1 above)
Comment (erikerikson) [inline lib/ProviderAws.js]: You are absolutely correct.  Thank you very much for catching this!

**Comment by erikerikson** [comment]
Please note that these changes changed environment variable patterns from:
`AWS_PROFILE_DEV`
to:
`AWS_DEV_PROFILE`
This enabled the use of the built in AWS.EnvironmentCredentials class.

I mention since this may have an impact on documentation.

**Comment by srg-clearlaunch** [comment]
:+1: we need this too

**Comment by ac360** [comment]
Thanks @erikerikson!  This looks great.  I'll review and get this in today!

**Comment by erikerikson** [comment]
@ac360 finally getting to run this through on a plausible project.  getting it fixed up, apologies.

**Comment by ac360** [comment]
@erikerikson Should we hold back the review?

**Comment by erikerikson** [comment]
Should be good now but doing a final verification

**Comment by erikerikson** [comment]
Verification is complete locally but the proxy configuration doesn't appear to have taken so I'm blocked on verifying the complete workflow for the moment.  I'm checking into that.

**Comment by erikerikson** [comment]
Looks like the proxy settings were being wiped out in ProviderAws.js on line 67 because they weren't included in the credentials bundle.  Investigating a fix.  Prefer that added here or shall I go with the more clean separated fix?

**Comment by erikerikson** [comment]
Okie, I used a bad environment variable.  Apologies.  My profile credentials were loaded and used correctly.  This should be good to go.

**Pull request merged**